### PR TITLE
install.sh: Improve stability of `curl`, disallow crostini/beta/canary channel

### PR DESCRIPTION
**Pull request by supechicken**
Merged from #5954 and #5967 

### Changes
---
- Added a retry loop for `curl` (the `--retry 3 --retry-all-errors` parameter does not work with the `Couldn't connect to server ` error)

Examples:
```
Downloading ca_certificates...
curl: (7) Couldn't connect to server                                          

Retrying, 3 retries left.
```
```
Downloading gmp...
curl: (6) Could not resolve host: gitlab.com
Retrying, 3 retries left.
curl: (6) Could not resolve host: gitlab.com
Retrying, 2 retries left.
curl: (6) Could not resolve host: gitlab.com
Retrying, 1 retries left.
curl: (6) Could not resolve host: gitlab.com
Retrying, 0 retries left.
Download failed :/ Please check your network settings.
```
---
- Force `curl` to use system libraries (`unset LD_LIBRARY_PATH`), use `ssl`
- Enable parallel download on `curl` if available
- Optimize the `url/sha256` extract logic
---
- Disallow Linux containers (`crostini`) and devices on `beta/dev/canary` update channels
```
$ curl -Ls https://raw.githubusercontent.com/supechicken666/chromebrew/install_sh/install.sh | bash
Crostini containers are not supported by Chromebrew :/
Run 'curl -Ls git.io/vddgY | CREW_FORCE_INSTALL=1 bash' to perform install anyway
```
---
Tested on `x86_64`

### Run the following to get this pull request's changes locally for testing.
```
curl -Ls https://raw.githubusercontent.com/supechicken666/chromebrew/install_sh/install.sh | bash
```

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -30,16 +30,61 @@ RED='\e[1;91m';    # Use Light Red for errors.
...
+  if [[ ! "$(< /etc/lsb-release)" =~ CHROMEOS_RELEASE_TRACK=stable-channel$'\n' && "${CREW_FORCE_INSTALL}" != '1' ]]; then
+    echo -e "${YELLOW}The beta, dev, and canary channel are unsupported by Chromebrew${RESET}"
+    echo -e "${YELLOW}Run 'curl -Ls git.io/vddgY | CREW_FORCE_INSTALL=1 bash' to perform install anyway${RESET}"
+    exit 1
+  fi
+else
+  echo -e "${YELLOW}Unable to detect system information, installation will continue.${RESET}"
+fi
+
+echo -e "\n${GREEN}Welcome to Chromebrew!${RESET}\n"
```
Comment (uberhacker) [inline install.sh:58]: This line should probably go beneath lines 59-62 below.

File: install.sh [hunk 2]
```diff
@@ -30,16 +30,61 @@ RED='\e[1;91m';    # Use Light Red for errors.
...
+if [[ -d /opt/google/cros-containers && "${CREW_FORCE_INSTALL}" != '1' ]]; then
+  echo -e "${RED}Crostini containers are not supported by Chromebrew :/${RESET}"
+  echo -e "${YELLOW}Run 'curl -Ls git.io/vddgY | CREW_FORCE_INSTALL=1 bash' to perform install anyway${RESET}"
+  exit 1
+fi
+
+# disallow non-stable channels Chrome OS
+if [ -f /etc/lsb-release ]; then
+  if [[ ! "$(< /etc/lsb-release)" =~ CHROMEOS_RELEASE_TRACK=stable-channel$'\n' && "${CREW_FORCE_INSTALL}" != '1' ]]; then
+    echo -e "${YELLOW}The beta, dev, and canary channel are unsupported by Chromebrew${RESET}"
```
Comment (uberhacker) [inline install.sh:50]: `{RED}` instead of `{YELLOW}`?

File: install.sh [hunk 3]
```diff
@@ -30,16 +30,61 @@ RED='\e[1;91m';    # Use Light Red for errors.
...
   exit 1;
 fi
 
+# prompt user to enter the sudo password if it set
+# if the PASSWD_FILE specified by chromeos-setdevpasswd exist, that means a sudo password is set
+if [[ "$(< /usr/sbin/chromeos-setdevpasswd)" =~ PASSWD_FILE=\'([^\']+) ]] && [ -f "${BASH_REMATCH[1]}" ]; then
+  echo -e "${BLUE}Please enter the developer mode password${RESET}"
+  # reset sudo timeout
+  sudo -k
+  sudo /bin/true
```
Comment (uberhacker) [inline install.sh:70]: Are 2 sudo lines necessary?

File: install.sh [hunk 4]
```diff
@@ -30,16 +30,61 @@ RED='\e[1;91m';    # Use Light Red for errors.
...
+# force curl to use system libraries
+function curl () {
+  # retry if download failed
+  # the --retry/--retry-all-errors parameter in curl will not work with the 'curl: (7) Couldn't connect to server'
+  # error, a for loop is used here
+  for (( i = 0; i < 4; i++ )); do
+    env LD_LIBRARY_PATH='' ${CURL} --ssl -C - "${@}" && \
+      return 0 || \
+      echo -e "${YELLOW}Retrying, $((3-$i)) retries left.${RESET}"
+  done
```
Comment (uberhacker) [inline install.sh:82]: Just so you know, this will be 4 retries instead of 3.

File: install.sh [hunk 5]
```diff
@@ -30,16 +30,61 @@ RED='\e[1;91m';    # Use Light Red for errors.
...
+  # the download failed if we still here
```
Comment (uberhacker) [inline install.sh:83]: `if we're still here`?

File: install.sh (hunk 3 above)
Comment (supechicken) [reply install.sh:70]: The first `sudo` is used to reset the password timeout, so we can confirm that a password prompt will be shown in next sudo command

A password prompt will show in the second `sudo`, then the following `sudo` will not ask for the password (unless the `sudo` password timestamp expired)

File: install.sh [hunk 6]
```diff
@@ -64,11 +109,9 @@ do
...
 sudo chown "$(id -u)":"$(id -g)" "${CREW_PREFIX}"
 
-# Delete 'var' symlink on Cloudready platform
-if [[ $(grep neverware /etc/lsb-release) != "" ]]; then
-  [ -L /usr/local/var ] && sudo rm -f /usr/local/var
-  [ -L /usr/local/local ] && sudo rm -f /usr/local/local
-fi
+# Delete ${CREW_PREFIX}/{var,local} symlink on some Chromium OS distro if exist
+[ -L ${CREW_PREFIX}/var ] && sudo rm -f #{CREW_PREFIX}/var
+[ -L ${CREW_PREFIX}/local ] && sudo rm -f #{CREW_PREFIX}/local
```
Comment (uberhacker) [inline install.sh:114]: Probably should enclose in double quotes in case `#{CREW_PREFIX}` has a space.  Not that it ever should happen but it could potentially lead to the wrong files being deleted depending on the directory name.

File: install.sh (hunk 4 above)
Comment (supechicken) [reply install.sh:82]: It means 1 initial try, then 3 retries😅

File: install.sh [hunk 7]
```diff
@@ -77,21 +120,25 @@ for dir in "${CREW_CONFIG_PATH}/meta" "${CREW_DEST_DIR}" "${CREW_PACKAGES_PATH}"
   fi
 done
 
+echo -e "\n${YELLOW}Downloading information for Bootstrap packages...${RESET}"
+echo -en "${GRAY}"
```
Comment (uberhacker) [inline install.sh:124]: Is this line necessary?

File: install.sh [hunk 8]
```diff
@@ -77,21 +120,25 @@ for dir in "${CREW_CONFIG_PATH}/meta" "${CREW_DEST_DIR}" "${CREW_PACKAGES_PATH}"
...
+# use parallel mode if available
+if [[ "$(curl --help curl)" =~ --parallel ]]; then
```
Comment (uberhacker) [inline install.sh:126]: `$(${CURL}...`?

File: install.sh [hunk 9]
```diff
@@ -136,7 +180,7 @@ function download_check () {
     fi
     #download
     echo -e "${BLUE}Downloading ${1}...${RESET}"
-    $CURL '-#' -L "${2}" -o "${3}"
+    curl '-#' -L "${2}" -o "${3}"
```
Comment (uberhacker) [inline install.sh:183]: Why was this line changed?  Don't we want to use the environment variable instead?

File: install.sh [hunk 10]
```diff
@@ -192,8 +236,8 @@ for i in $(seq 0 $((${#urls[@]} - 1))); do
   tarfile="$(basename ${url})"
   name="${tarfile%%-*}"   # extract string before first '-'
   rest="${tarfile#*-}"    # extract string after first '-'
-  version="$(echo ${rest} | sed -e 's/-chromeos.*$//')"
-                        # extract string between first '-' and "-chromeos"
+  version="${rest%%-chromeos*}"
```
Comment (uberhacker) [inline install.sh:239]: Good call.

File: install.sh (hunk 9 above)
Comment (supechicken) [reply install.sh:183]: The `${CURL}` variable is already applied to the `curl` command in `function curl()`:
https://github.com/skycocker/chromebrew/blob/ecfc304e46e082c5f2490f097f5f2acf1e86977c/install.sh#L74-L79

File: install.sh (hunk 8 above)
Comment (supechicken) [reply install.sh:126]: See https://github.com/skycocker/chromebrew/pull/6519#discussion_r775123014

File: install.sh (hunk 7 above)
Comment (supechicken) [reply install.sh:124]: It is used to set the output color of `curl`

File: install.sh (hunk 2 above)
Comment (supechicken) [reply install.sh]: Changed

File: install.sh (hunk 5 above)
Comment (supechicken) [reply install.sh]: Done

File: install.sh (hunk 4 above)
Comment (uberhacker) [reply install.sh:83]: The initial try is still a retry. ;)

Comment (uberhacker) [reply install.sh:83]: If we're defining a curl function here, why do we need the `CURL` environment variable?

Comment (uberhacker) [reply install.sh:83]: OK, I think I understand what is going on here.  The `CURL` environment variable is used inside this function.

File: install.sh [hunk 11]
```diff
@@ -104,18 +152,15 @@ if [ ! -f device.json ]; then
...
-  temp_sha256="$(echo "$pkgfile" | grep -m 3 "$ARCH": | tail -n 1 | cut -d\' -f2 | tr -d \' | tr -d \" | sed 's/,//g')"
-  urls[k]="$temp_url"
-  sha256s[k]="$temp_sha256"
-  k=$((k+1))
+  pkgfile="${CREW_PACKAGES_PATH}/${package}.rb"
+
+  [[ "$(sed -n '/binary_sha256/,/}/p' "${pkgfile}")" =~ .*x86_64:[[:blank:]]*[\'\"]([^\'\"]*) ]]
+    sha256s+=("${BASH_REMATCH[1]}")
+
+  [[ "$(sed -n '/binary_url/,/}/p' "${pkgfile}")" =~ .*x86_64:[[:blank:]]*[\'\"]([^\'\"]*) ]]
```
Comment (uberhacker) [inline install.sh:162]: Shouldn't we abstract the arch with `$ARCH` instead of `x86_64`?

Comment (supechicken) [reply install.sh]: Change made
